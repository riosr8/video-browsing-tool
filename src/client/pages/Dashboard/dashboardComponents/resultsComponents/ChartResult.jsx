import React from 'react';
import PropTypes from 'prop-types';
import Chart from 'react-google-charts';

const ChartResult = props => {
  const { data, chartType } = props;
  if (chartType === '2Ddata') {
    return (
      <Chart
        chartType="LineChart"
        loader={<div>Loading Chart</div>}
        data={data}
        options={{
          hAxis: {
            title: 'Threshold',
          },
          vAxis: {
            title: 'Correlation',
          },
          series: {
            1: { curveType: 'function' },
          },
          legend: {
            position: 'none',
          },
        }}
      />
    );
  }
  if (chartType === 'histogram') {
    return (
      <Chart
        chartType="Bar"
        loader={<div>Loading Chart</div>}
        data={data}
        options={{
          // Material design options
          legend: {
            position: 'top',
          },
        }}
      />
    );
  }
};

export default ChartResult;

ChartResult.propTypes = {
  data: PropTypes.array.isRequired,
  chartType: PropTypes.string.isRequired,
};
