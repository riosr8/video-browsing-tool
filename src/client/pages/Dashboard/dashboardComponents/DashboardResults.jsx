import React, { useState } from 'react';
import {
  Transition,
  Button,
  Icon,
  Grid,
  Dropdown,
  Segment,
  Header,
  Container,
} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import ReactPlayer from 'react-player';
import VideoResult from './resultsComponents/VideoResult';
import ChartResult from './resultsComponents/ChartResult';

const DashboardResults = props => {
  const { resultsIsVisible, setResultsIsVisible, fetchingResults, results } = props;

  if (!resultsIsVisible) {
    return null;
  }

  if (fetchingResults) {
    return (
      <Transition.Group animation="slide down" duration="300">
        <Container style={{ marginTop: '2%' }}>
          <Segment loading placeholder />
        </Container>
      </Transition.Group>
    );
  }

  if (results.results.length === 0) {
    return (
      <Transition.Group animation="slide down" duration="300">
        <Container style={{ marginTop: '2%' }}>
          <Button
            id="closeVideoPlayer"
            icon
            color="red"
            onClick={() => {
              setResultsIsVisible(false);
            }}
          >
            <Icon name="close" size="small" />
          </Button>
          <Segment placeholder>
            <Header
              icon
              content="No results available. Please submit a processing request first."
            />
          </Segment>
        </Container>
      </Transition.Group>
    );
  }
  const [selectedResult, setSelectedResult] = useState(results.results[0].value);

  return (
    <Transition.Group animation="slide down" duration="300">
      {resultsIsVisible && (
        <Grid stackable centered style={{ marginLeft: '2%', marginRight: '2%', marginTop: '2%' }}>
          <Grid.Row columns={2}>
            <Grid.Column>
              <Button
                id="closeVideoPlayer"
                icon
                color="red"
                onClick={() => {
                  setResultsIsVisible(false);
                }}
              >
                <Icon name="close" size="small" />
              </Button>
            </Grid.Column>
            <Grid.Column>
              <Dropdown
                placeholder="Choose an algorithm"
                fluid
                selection
                value={selectedResult}
                options={results.results}
                onChange={(e, data) => {
                  e.preventDefault();
                  setSelectedResult(data.value);
                }}
              />
            </Grid.Column>
          </Grid.Row>
          <Grid.Row columns={2}>
            <Grid.Column>
              <div
                style={{
                  paddingTop: '56.25%',
                  position: 'relative',
                }}
              >
                <ReactPlayer
                  style={{
                    position: 'absolute',
                    top: 0,
                    bottom: 0,
                  }}
                  url={`http://localhost:5000/noAuth/videos/${results.videoUuid}`}
                  controls
                  width="100%"
                  height="100%"
                />
              </div>
            </Grid.Column>
            <Grid.Column>
              {results.results
                .filter(algo => algo.id === selectedResult)
                .map(algo => {
                  switch (algo.type) {
                    case 'video':
                      return <VideoResult key={algo.id} url={algo.result} />;
                    case '2Ddata':
                      return <ChartResult key={algo.id} data={algo.result} chartType="2Ddata" />;
                    case 'histogram':
                      return <ChartResult key={algo.id} data={algo.result} chartType="histogram" />;
                    default:
                      return null;
                  }
                })}
            </Grid.Column>
          </Grid.Row>
        </Grid>
      )}
    </Transition.Group>
  );
};

DashboardResults.propTypes = {
  resultsIsVisible: PropTypes.bool.isRequired,
  setResultsIsVisible: PropTypes.func.isRequired,
  fetchingResults: PropTypes.bool.isRequired,
  results: PropTypes.object,
};

export default DashboardResults;
