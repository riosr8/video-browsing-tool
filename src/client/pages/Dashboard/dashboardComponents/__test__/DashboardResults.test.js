import React from 'react';
import { mount } from 'enzyme';
import { Segment, Header, Dropdown } from 'semantic-ui-react';
import { act } from 'react-dom/test-utils';
import DashboardResults from '../DashboardResults';
import mockResults from '../mock-data/mockData';
import ChartResult from '../resultsComponents/ChartResult';

describe('Dashboard Results', () => {
  let wrapper;
  const mockSetResultsIsVisible = jest.fn();

  beforeEach(() => {
    wrapper = mount(
      <DashboardResults
        results={mockResults}
        resultsIsVisible
        setResultsIsVisible={mockSetResultsIsVisible}
        fetchingResults={false}
      />,
    );
  });

  it('renders with out crashing ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should close results view on close click', () => {
    wrapper
      .find('#closeVideoPlayer')
      .hostNodes()
      .simulate('click');
    expect(mockSetResultsIsVisible.mock.calls).toHaveLength(1);
  });

  it('should trigger dropdown onChange function', () => {
    act(() => {
      wrapper.find(Dropdown).prop('onChange')({ preventDefault: jest.fn() }, { value: '1' });
    });

    expect(wrapper.find(Dropdown).props().value).toBe('1');
  });

  it('should render a chart given 2D data', () => {
    wrapper.setProps({
      results: {
        videoUuid: '19616905-83ff-4671-8c02-7c62d13e2458big_buck_bunny.mp4',
        results: [
          {
            key: 'Algorithm 1',
            value: '1',
            text: 'Algorithm 1',
            id: '1',
            result: [],
            type: '2Ddata',
          },
        ],
      },
    });

    expect(wrapper.find(ChartResult)).toHaveLength(1);
  });

  it('should render a chart given histogram data', () => {
    wrapper.setProps({
      results: {
        videoUuid: '19616905-83ff-4671-8c02-7c62d13e2458big_buck_bunny.mp4',
        results: [
          {
            key: 'Algorithm 1',
            value: '1',
            text: 'Algorithm 1',
            id: '1',
            result: [],
            type: 'histogram',
          },
        ],
      },
    });

    expect(wrapper.find(ChartResult)).toHaveLength(1);
  });

  it('should return null if chart type is not handled', () => {
    wrapper.setProps({
      results: {
        videoUuid: '19616905-83ff-4671-8c02-7c62d13e2458big_buck_bunny.mp4',
        results: [
          {
            key: 'Algorithm 1',
            value: '1',
            text: 'Algorithm 1',
            id: '1',
            result: [],
            type: '',
          },
        ],
      },
    });

    expect(wrapper.find(ChartResult)).toHaveLength(0);
  });

  it('should return segment in loading state', () => {
    wrapper.setProps({ fetchingResults: true });
    expect(wrapper.find(Segment).findWhere(n => n.prop('loading') === true)).toHaveLength(1);
  });

  it('should return a header with message if no results', () => {
    wrapper.setProps({ results: { videoUuid: 'id', results: [] } });
    expect(
      wrapper.find(Header).findWhere(n => {
        return (
          n.prop('content') === 'No results available. Please submit a processing request first.'
        );
      }),
    ).toHaveLength(1);
  });

  it('should close header with message if no results on close click', () => {
    wrapper.setProps({ results: { videoUuid: 'id', results: [] } });
    wrapper
      .find('#closeVideoPlayer')
      .hostNodes()
      .simulate('click');
    expect(mockSetResultsIsVisible.mock.calls).toHaveLength(1);
  });
});
