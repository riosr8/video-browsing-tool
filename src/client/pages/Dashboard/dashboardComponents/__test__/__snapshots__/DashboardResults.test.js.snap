// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Results renders with out crashing  1`] = `ReactWrapper {}`;
